napari/components/viewer_model.py

```diff
@@ -914,17 +908,139 @@ def open(
+        are available to read these paths, an error is returned specifying
+        this.
+
+        Errors *are* raised by this function, when the given paths are not
+        a list or tuple, or if no plugins are available to read the files.
+        This assumes all files have the same extension, as other cases
+        are not yet supported.
+
+        This function is called from ViewerModel.open, which raises any
+        errors returned. The QtViewer also calls this method but handles
+        errors by raising a dialog for users to make a plugin choice.
+
+        Parameters
+        ----------
+        paths : List[Path | str]
+            list of file paths to open
+        kwargs : Dict[str, Any], optional
+            keyword arguments to pass to layer adding method, by default {}
+        layer_type : Optional[str], optional
+            layer type for paths, by default None
+        stack : bool, optional
+            True if files should be opened as a stack, by default False
+
+        Returns
+        -------
+        added
+            list of layers added
+        plugin
+            plugin used to try opening paths, if any
+        error
+            errors from trying to open layers
+
+        Raises
+        ------
+        ValueError
+            when paths is *not* a list or tuple, or no plugins are available
+        """
+        paths = [os.fspath(path) for path in paths]  # PathObjects -> str
+        if not isinstance(paths, (tuple, list)):
+            raise ValueError(
+                trans._(
+                    "'path' argument must be a string, list, or tuple",
+                    deferred=True,
+                )
+            )
+
+        added = []
+        error = None
+        plugin = None
+        _path = paths[0]
+
+        readers = get_potential_readers(_path)
+        if not readers:
+            raise ValueError(
+                trans._(
+                    'No plugin found capable of reading {paths}.',
+                    deferred=True,
+                    paths=paths,
+                )
+            )
+
+        plugin = get_preferred_reader(_path)
+
+        # TODO: Use custom error instance?
+        # preferred plugin exists, or we just have one plugin available
+        if plugin in readers or (not plugin and len(readers) == 1):
+            plugin = plugin or next(iter(readers.keys()))
+            try:
+                added = self._add_layers_with_plugins(
+                    paths,
+                    kwargs=kwargs,
+                    stack=stack,
+                    plugin=plugin,
+                    layer_type=layer_type,
+                )
+            # plugin failed
+            except Exception as e:
+                # TODO: we're changing the error type here and we shouldn't maybe?
+                error = RuntimeError(
+                    f'Tried opening with reader {plugin}, but failed:\n'
+                    + str(e)
+                )
```

Comment:
@DragaDoncila and I just discussed this in a call. My suggestion is that passing error instances up the call stack is rather unpythonic, but, worse, this masks the traceback from the original exception, which would be very useful to someone debugging. Therefore, @DragaDoncila will try to rewrite this using `raise RuntimeError(...) from e`. Thanks @DragaDoncila!

That's the only minor issue I have with the PR.

Reply:
Noted @jni will ping when this is ready for re-review. It's going to be Monday Australia time